feat: add more reservation control to create nexus

tiagolobocastro opened the pull request:
This allows the create to specify whether we want to preempt the existing holder when the existing holder is not known.
EG: If a nexus create times out, we don't know if the reservations have changed or not, so if we attempt to create a nexus, which reservation key do we preempt? With `NexusNvmePreemption::Holder` we can preempt the existing holder, whomever that might be.

This logic is already kind used by the dataplane, but only if the existing holder is type 1, which we cannot create at the moment as we hardcode to 5, which is a "shared" scheme. By passing the `resvType` as an arg, it allows us to control that, and also might be useful for the future.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat: add more reservation control to create nexus

This allows the create to specify whether we want to preempt the existing holder when the
existing holder is not known.
EG: If a nexus create times out, we don't know if the reservations have changed or not, so
if we attempt to create a nexus, which reservation key do we preempt?
With `NexusNvmePreemption::Holder` we can preempt the existing holder, whomever that might be.

This logic is already kind used by the dataplane, but only if the existing holder is type 1,
which we cannot create at the moment as we hardcode to 5, which is a "shared" scheme.
By passing the `resvType` as an arg, it allows us to control that, and also might be useful
for the future.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

mtzaurus commented on io-engine/src/bdev/nexus/nexus_bdev.rs:
```diff
@@ -99,17 +100,60 @@ impl NvmeAnaState {
+            4 => Self::ExclusiveAccessRegsOnly,
+            5 => Self::WriteExclusiveAllRegs,
+            6 => Self::ExclusiveAccessAllRegs,
+            _ => return Err(()),
```
Shall we return a string error here that provides details ?

tiagolobocastro replied:
Yep, we should indeed.

tiagolobocastro replied:
Done

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors try

tiagolobocastro committed:
test(nvme): sort the container handles

Otherwise tests might yield surprising results, as we'd be calling the wrong instance in
 tests where we use arrays indexes..

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
bors try

tiagolobocastro commented:
Seems the rebuild grpc failure is still haunting us
bors try

tiagolobocastro committed:
ci: fix nodejs test by using retry function

Use the retry patter with a high timeout instead of simply sleeping..

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors merge
